src/webcore/executor.rs

```diff
@@ -0,0 +1,194 @@
+use futures::future::{Future, ExecuteError, Executor};
+use futures::executor::{self, Notify, Spawn};
+use futures::Async;
+use std::collections::VecDeque;
+use std::result::Result as StdResult;
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
+use webcore::try_from::TryInto;
+use webcore::value::Reference;
+
+
+const INITIAL_QUEUE_CAPACITY: usize = 10;
+
+
+// This functionality should really be in libstd, because the implementation
+// looks stupid.
+unsafe fn clone_raw< T >( ptr: *const T ) -> Rc< T > {
+    let result = Rc::from_raw( ptr );
+    ::std::mem::forget( result.clone() );
+    result
+}
+
+// Typing this out is tedious
+type BoxedFuture = Box< Future< Item = (), Error = () > + 'static >;
+
+struct SpawnedTask {
+    is_queued: Cell< bool >,
+    spawn: RefCell< Option< Spawn< BoxedFuture > > >,
+}
+
+impl SpawnedTask {
+    fn new< F >( future: F ) -> Rc< Self >
+        where F: Future< Item = (), Error = () > + 'static {
+        Rc::new( Self {
+            is_queued: Cell::new( false ),
+            spawn: RefCell::new( Some( executor::spawn(
+                Box::new( future ) as BoxedFuture
+            ) ) )
+        } )
+    }
+
+    fn poll( &self ) {
+        let mut spawn = self.spawn.borrow_mut();
+
+        // Take the future so that if we panic it gets dropped
+        if let Some( mut spawn_future ) = spawn.take() {
+            // Clear `is_queued` flag so that it will re-queue if poll calls task.notify()
+            self.is_queued.set( false );
+
+            if spawn_future.poll_future_notify( &&EventLoop, self as *const _ as usize ) == Ok( Async::NotReady ) {
+                // Future was not ready, so put it back
+                *spawn = Some( spawn_future );
+            }
+        }
+    }
+
+    fn notify( task: Rc< SpawnedTask > ) {
+        // If not already queued
+        if !task.is_queued.replace( true ) {
+            EventLoop.push_task(task);
+        }
+    }
+}
+
+// A proxy for the javascript event loop.
+struct EventLoop;
+
+// There's only one thread, but this lets us tell the compiler that we
```

Comment:
This note doesn't seem correct. `thread_local` has nothing to do with "telling the compiler that we don't need a `Sync` bound".

Instead, it doesn't need a `Sync` bound because it will actually create a new value on each thread, so if there's 5 threads it might create 5 `EventLoopInner`

Reply:
In the absence of multiple threads, a "thread-local" is a way of telling the compiler that globals do not need to be thread-safe. (ie. not `Sync`) whilst also achieving lazy initialisation, so I think the comment is pretty accurate.